rcsb/utils/chemref/RcsbLigandReferenceGenerator.py

```diff
@@ -0,0 +1,300 @@
+from sklearn.preprocessing import StandardScaler
+from rcsbapi.data import DataQuery as Query
+from rcsbapi.data import ALL_STRUCTURES
+from rcsbapi.config import config
+from rcsb.utils.io.StashableBase import StashableBase
+
+logger = logging.getLogger(__name__)
+
+
+class RcsbLigandReferenceGenerator(StashableBase):
+    """ This class generates ligand quality reference data by performing the following steps:
+    Query ligand quality metrics from RCSB GraphQL API;
+    Process the ligand quality data by filtering, aggregating, and formatting;
+    Generate ligand reference data through PCA;
+        Attributes:
+            data: Stateful data updated with each step
+        Methods:
+            query(pdb_id): Fetch ligand quality metrics for given PDB IDs.
+            filter(): Filter the queried ligand quality scores based on defined criteria.
+            reduce(): Reduce the data by combining multiple instances of the same ligand in the same PDB entry.
+            analyze(): Run PCA on the filtered and reduced ligand quality scores to generate reference data.
+            generate(pdb_ids): Full pipeline to generate ligand quality reference data by running the steps of
+                query -> filter -> reduce -> analyze.
+            writeReference(output_file): Write the generated ligand quality reference data to a csv file.
+    """
+    def __init__(self):
+        self.data = None
+        # Increase some settings to speed up querying
+        config.DATA_API_MAX_CONCURRENT_REQUESTS = 15
+        config.DATA_API_REQUESTS_PER_SECOND = 30
+
+    def generate(self, pdb_ids: list[str] = []) -> "RcsbLigandReferenceGenerator":
+        """
+        Full pipeline to generate ligand quality reference data by running the steps of
+        query -> filter -> reduce -> analyze.
+
+        :param pdb_ids: List of specified PDB IDs, default to [] which leads to all PDB structures being queried.
+        :return: The same object with updated self.data of ligand quality reference data.
+        """
+        self.query(pdb_ids).filter().reduce().analyze()
+        return self
+
+    def writeReference(self, output_file: str) -> bool:
+        """
+        Write the generated ligand quality reference data to a csv file.
+
+        :param output_file: Path to the output csv file.
+        :return: True upon successful write operation.
+        """
+        if not self.data:
+            logger.error("No data to write. Please run generate() first.")
+            return False
+        if type(self.data) is not list:
+            logger.error("Data format incorrect. Expected a list of dictionaries after generate()")
+            return False
+        if type(self.data[0]) is not dict:
+            logger.error("Data format incorrect. Expected a list of dictionaries after generate()")
+            return False
+        fieldnames = self.data[0].keys()
+        with open(output_file, mode="w", newline="") as file:
```

Comment:
The linter also wants an encoding specified here:

```suggestion
        with open(output_file, mode="w", newline="", encoding="utf-8") as file:
```